// node-env-run/dist/lib/cli.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var path = require("path");
var yargs = require("yargs");
var dotenv = require("dotenv");
var Debug = require("debug");
var pkginfo = require("pkginfo");
var common_tags_1 = require("common-tags");
var utils_1 = require("./utils");
var debug = Debug('node-env-run');
var cwd = process.cwd();
var pkg = pkginfo(module);
var usageDescription = common_tags_1.stripIndent(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  Runs the given script with set environment variables. \n  * If no script is passed it will run the REPL instead.\n  * If '.' is passed it will read the package.json and execute the 'main' file.\n\n  Pass additional arguments to the script after the \"--\"\n"], ["\n  Runs the given script with set environment variables. \n  * If no script is passed it will run the REPL instead.\n  * If '.' is passed it will read the package.json and execute the 'main' file.\n\n  Pass additional arguments to the script after the \"--\"\n"])));
function parseArgs(argv) {
    var result = yargs
        .usage('$0 [script]', usageDescription, function (yargs) {
        yargs.positional('script', {
            describe: 'the file that should be executed',
            type: 'string',
        });
        yargs.example('$0 --exec "python"', 'Runs the Python REPL instead');
        yargs.example('$0 server.js --exec "nodemon"', 'Runs "nodemon server.js"');
        yargs.example('$0 someScript.js -- --inspect', 'Run script with --inspect');
        return yargs;
    })
        .option('force', {
        alias: 'f',
        demandOption: false,
        describe: 'temporarily overrides existing env variables with the ones in the .env file',
    })
        .option('env', {
        alias: 'E',
        demandOption: false,
        describe: 'location of .env file relative from the current working directory',
        default: '.env',
        type: 'string',
    })
        .option('verbose', {
        demandOption: false,
        describe: 'enable verbose logging',
        type: 'boolean',
    })
        .option('encoding', {
        demandOption: false,
        describe: 'encoding of the .env file',
        default: 'utf8',
        type: 'string',
    })
        .option('exec', {
        alias: 'e',
        demandOption: false,
        describe: 'the command to execute the script with',
        default: 'node',
    })
        .showHelpOnFail(true)
        .help('help')
        .strict()
        .version()
        .parse(argv.slice(2));
    var script = result.script;
    result.newArguments = result._;
    return { program: result, script: script };
}
exports.parseArgs = parseArgs;
function init(args) {
    var program = args.program, script = args.script;
    var envFilePath = path.resolve(cwd, program.env);
    if (!fs.existsSync(envFilePath)) {
        var error = new Error("Could not find the .env file under: \"" + envFilePath + "\"");
        return { isRepl: false, error: error };
    }
    debug('Reading .env file');
    var envContent = fs.readFileSync(path.resolve(cwd, program.env), program.encoding);
    var envValues = dotenv.parse(envContent);
    utils_1.setEnvironmentVariables(envValues, program.force);
    if (!script) {
        var node = args.program.exec === undefined;
        return { isRepl: true, node: node };
    }
    var scriptToExecute = utils_1.getScriptToExecute(script, cwd);
    if (scriptToExecute === null || !fs.existsSync(scriptToExecute)) {
        var error = new Error('Failed to determine script to execute');
        return { isRepl: false, error: error };
    }
    return { isRepl: false, script: scriptToExecute };
}
exports.init = init;
var templateObject_1;
